bpf/bpf_sock.c

```diff
@@ -273,27 +276,60 @@ static __always_inline int __sock4_xlate(struct bpf_sock_addr *ctx,
 	if (sock4_skip_xlate(svc, in_hostns, ctx->user_ip4))
 		return -EPERM;
 
-	key.slave = (sock_local_cookie(ctx_full) % svc->count) + 1;
+#if defined(ENABLE_SESSION_AFFINITY) && defined(BPF_HAVE_NETNS_COOKIE)
+	/* Session affinity id (a netns cookie) */
+	__u64 client_id = 0;
+	if (svc->affinity) {
+		client_id = get_netns_cookie(ctx_full);
+		backend_id = lb4_affinity_backend_id(svc->rev_nat_index,
+						     svc->affinity_timeout,
+						     true, client_id);
+		backend_from_affinity = true;
+	}
+#endif
 
-	slave_svc = __lb4_lookup_slave(&key);
-	if (!slave_svc) {
-		update_metrics(0, METRIC_EGRESS, REASON_LB_NO_SLAVE);
-		return -ENOENT;
+	if (backend_id == 0) {
+reselect_backend: __maybe_unused
+		backend_from_affinity = false;
+		key.slave = (sock_local_cookie(ctx_full) % svc->count) + 1;
+		slave_svc = __lb4_lookup_slave(&key);
+		if (!slave_svc) {
+			update_metrics(0, METRIC_EGRESS, REASON_LB_NO_SLAVE);
+			return -ENOENT;
+		}
+		backend_id = slave_svc->backend_id;
 	}
 
-	backend = __lb4_lookup_backend(slave_svc->backend_id);
+	backend = __lb4_lookup_backend(backend_id);
 	if (!backend) {
+#if defined(ENABLE_SESSION_AFFINITY) && defined(BPF_HAVE_NETNS_COOKIE)
+		if (backend_from_affinity) {
+			/* Backend from the session affinity no longer exists,
+			 * thus select a new one. Also, remove the affinity,
+			 * so that if the svc doesn't have any backend,
+			 * a subsequent request to the svc doesn't hit
+			 * the reselection again. */
+			lb4_delete_affinity(svc->rev_nat_index, true, client_id);
+			goto reselect_backend;
+		}
+#endif
```

Comment:
What happens if backend_id was reused for a different service in the meantime, there is no detection here right now?

Reply:
Do you mean a case when `goto reselect_backend` happens? In this case, a new backend will be selected via `__lb4_lookup_slave()`. In the worst case (the selected backend no longer exists due to invalid `svc->count`), a request will be dropped or handled by bpf_lxc.

Or do you mean in general? If so, `lb{4,6}_affinity_backend_id()` does a lookup in the `LB_AFFINITY_MATCH_MAP` (key = svc ID + backend ID) to see whether the backend ID is still used by the svc.